pkg/ingester/activeseries/active_series.go

```diff
@@ -200,34 +225,77 @@ func (s *seriesStripe) updateSeriesTimestamp(now time.Time, series labels.Labels
 	}
 }
 
-func (s *seriesStripe) findEntryForSeries(ref uint64) *atomic.Int64 {
+func (s *seriesStripe) findEntryForSeries(ref uint64, hasNativeHistograms bool, numNativeHistogramBuckets int) (*atomic.Int64, bool) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	return s.refs[ref].nanos
+	entry := s.refs[ref]
+	return entry.nanos, entry.isNativeHistogram != hasNativeHistograms || entry.numNativeHistogramBuckets != numNativeHistogramBuckets
 }
 
-func (s *seriesStripe) findOrCreateEntryForSeries(ref uint64, series labels.Labels, nowNanos int64) (*atomic.Int64, bool) {
+func (s *seriesStripe) findAndUpdateOrCreateEntryForSeries(ref uint64, series labels.Labels, nowNanos int64, hasNativeHistograms bool, numNativeHistogramBuckets int) (*atomic.Int64, bool) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	matches := s.matchers.matches(series)
+	matchesLen := matches.len()
+
 	// Check if already exists within the entries.
 	// This repeats findEntryForSeries(), but under write lock.
 	entry, ok := s.refs[ref]
 	if ok {
+		if entry.isNativeHistogram && !hasNativeHistograms {
```

Comment:
if we had numNativeHistogramBuckets == -1 mean not a native histogram, then we can shortcut this code to one condition evaluation for float series: `entry.numNativeHistogramBuckets != numNativeHistogramBuckets` which would mean we have to deal with a type change or bucket count change, otherwise we just return without further tests since neither the type nor the bucket count changed.